src/utilities/oas-validator.ts

```diff
@@ -114,31 +114,103 @@ class OASValidator {
     path: string[],
   ): ValidationFailure[] {
     let failures: ValidationFailure[] = [];
+    const actualType = typeof actual;
+
+    const invalidSchemaErrors = this.checkInvalidSchemaErrors(
+      actual,
+      expected,
+      path,
+    );
+    if (invalidSchemaErrors) {
+      failures = [...failures, ...invalidSchemaErrors];
+      return failures;
+    }
+
+    const expectedTypeError = this.checkExpectedTypeError(
+      actual,
+      expected,
+      path,
+    );
+
+    if (expectedTypeError) {
+      failures = [...failures, expectedTypeError];
+      return failures;
+    }
+
+    const enumValueErrors = this.checkEnumValueError(actual, expected, path);
+    if (enumValueErrors.length > 0) {
+      failures = [...failures, ...enumValueErrors];
+    }
+
+    if (Array.isArray(actual)) {
+      const itemSchemaErrors = this.checkItemSchemaErrors(
+        actual,
+        expected,
+        path,
+      );
+      if (itemSchemaErrors) {
+        failures = [...failures, ...itemSchemaErrors];
+      }
+    } else if (actualType === 'object') {
+      const propertiesErrors = this.checkObjectPropertiesErrors(
+        actual,
+        expected,
+        path,
+      );
+      if (propertiesErrors) {
+        failures = [...failures, ...propertiesErrors];
+      }
+    }
+
+    return failures;
+  }
+
+  private static checkInvalidSchemaErrors(
```

Comment:
What do you think about just naming this `checkInvalidSchema`, since we're returning validation failures instead of throwing errors? Same thought for the other methods below.

Reply:
Looks like this method is only checking if null is allowed. Maybe we could name it something even more specific? Or do you all think we'll potentially add other checks in here later?